feat!: Add support for asm v1.8 to the asm module

kaariger opened the pull request:
- Modified default value for asm_version to 1.8
- Modified install_asm.sh wrapper script to call the new install_asm script with the updated flags. Removed flag to for the temp directory, allowing the install_asm script to manage its own temp directory.
    - TODO: Need to remove the --enable_cluster_labels flag in the wrapper script once the install_asm script makes that flag optional. It is currently required.
- Added additional IAM bindings and API enablement to the test scripts
- Regenerated docs
- Ran lint - only saw warnings about other files that are unmodified in this change
- Ran integration tests
- Fixes #707

morgante commented:
@kaariger What's the reason for a new PR instead of updating #821?

bharathkkb reviewed (commented):
Thanks for the PR @kaariger 

We can wait for upstream to remove the asmv validate as it will cause a diff on cluster labels if the cluster is managed by TF.

kaariger commented:
@morgante its a mistake on my part. I should have just updated my original PR. I closed the original PR (my mistake). Please let me know if I should update this PR to make it easier. For future, I will update existing PRs.

@bharathkkb it may be ok to wait. However, I had heard that some users were waiting for the module to support 1.8. Don't know if there is any urgent need to allow it as is with the labels added by install_asm script.

morgante commented:
> @morgante its a mistake on my part. I should have just updated my original PR. I closed the original PR (my mistake). Please let me know if I should update this PR to make it easier. For future, I will update existing PRs.

No worries, thanks! Just wanted to confirm if something prompted it.

comment-bot-dev commented:
Thanks for the PR! 🚀
Unfortunately it looks like some of our CI checks failed. See the [Contributing Guide](../blob/master/CONTRIBUTING.md) for details.
- ⚠️check_terraform
 Failed Terraform check. More details below. 
 ```bash 
Attempting to download /workspace/test/bundle.hcl bundle.
/tmp/bundler /workspace
Fetching Terraform 0.13.6 core package...
Local plugin directory ".plugins" found; scanning for provider binaries.
No ".plugins" directory found, skipping local provider discovery.
Creating terraform_0.13.6-bundle2021022300_linux_amd64.zip ...
All done!
Archive:  terraform_0.13.6-bundle2021022300_linux_amd64.zip
/workspace
Running terraform fmt
Running terraform validate
terraform_validate . 
Success!
 The configuration is valid.
terraform_validate ./examples/deploy_service 
Success!
 The configuration is valid.
terraform_validate ./examples/disable_client_cert 
Success!
 The configuration is valid.
terraform_validate ./examples/node_pool 
Success!
 The configuration is valid.
terraform_validate ./examples/node_pool_update_variant 
Success!
 The configuration is valid.
terraform_validate ./examples/node_pool_update_variant_beta 
Success!
 The configuration is valid.
terraform_validate ./examples/node_pool_update_variant_public_beta 
Success!
 The configuration is valid.
terraform_validate ./examples/private_zonal_with_networking 
Warning: 
Interpolation-only expressions are deprecated
  on main.tf line 42, in module "gcp-network":
  42:     
"${var.subnetwork}"
 = [
Terraform 0.11 and earlier required all non-constant expressions to be
provided via interpolation syntax, but this pattern is now deprecated. To
silence this warning, remove the "${ sequence from the start and the }"
sequence from the end of this expression, leaving just the inner expression.
Template interpolation syntax is still used to construct strings from
expressions when the template includes multiple interpolation sequences or a
mixture of literal strings and interpolations. This deprecation applies only
to templates that consist entirely of a single interpolation sequence.
Success!
 The configuration is valid, but there were some validation warnings as shown above.
terraform_validate ./examples/regional_private_node_pool_oauth_scopes 
Success!
 The configuration is valid.
terraform_validate ./examples/safer_cluster 
Warning: 
Interpolation-only expressions are deprecated
  on network.tf line 37, in module "gcp-network":
  37:     
"${local.subnet_name}"
 = [
Terraform 0.11 and earlier required all non-constant expressions to be
provided via interpolation syntax, but this pattern is now deprecated. To
silence this warning, remove the "${ sequence from the start and the }"
sequence from the end of this expression, leaving just the inner expression.
Template interpolation syntax is still used to construct strings from
expressions when the template includes multiple interpolation sequences or a
mixture of literal strings and interpolations. This deprecation applies only
to templates that consist entirely of a single interpolation sequence.
Success!
 The configuration is valid, but there were some validation warnings as shown above.
terraform_validate ./examples/safer_cluster_iap_bastion 
Warning: 
Interpolation-only expressions are deprecated
  on network.tf line 36, in module "vpc":
  36:     
"${var.subnet_name}"
 = [
Terraform 0.11 and earlier required all non-constant expressions to be
provided via interpolation syntax, but this pattern is now deprecated. To
silence this warning, remove the "${ sequence from the start and the }"
sequence from the end of this expression, leaving just the inner expression.
Template interpolation syntax is still used to construct strings from
expressions when the template includes multiple interpolation sequences or a
mixture of literal strings and interpolations. This deprecation applies only
to templates that consist entirely of a single interpolation sequence.
Error: 
Unsupported argument
  on .terraform/modules/cloud-nat/main.tf line 58, in resource "google_compute_router_nat" "main":
  58:   
enable_endpoint_independent_mapping
 = var.enable_endpoint_independent_mapping
An argument named "enable_endpoint_independent_mapping" is not expected here.
terraform_validate ./examples/shared_vpc 
Success!
 The configuration is valid.
terraform_validate ./examples/simple_regional 
Success!
 The configuration is valid.
terraform_validate ./examples/simple_regional_beta 
Success!
 The configuration is valid.
terraform_validate ./examples/simple_regional_private 
Success!
 The configuration is valid.
terraform_validate ./examples/simple_regional_private_beta 
Success!
 The configuration is valid.
terraform_validate ./examples/simple_regional_with_kubeconfig 
Success!
 The configuration is valid.
terraform_validate ./examples/simple_regional_with_networking 
Warning: 
Interpolation-only expressions are deprecated
  on main.tf line 45, in module "gcp-network":
  45:     
"${var.subnetwork}"
 = [
Terraform 0.11 and earlier required all non-constant expressions to be
provided via interpolation syntax, but this pattern is now deprecated. To
silence this warning, remove the "${ sequence from the start and the }"
sequence from the end of this expression, leaving just the inner expression.
Template interpolation syntax is still used to construct strings from
expressions when the template includes multiple interpolation sequences or a
mixture of literal strings and interpolations. This deprecation applies only
to templates that consist entirely of a single interpolation sequence.
Success!
 The configuration is valid, but there were some validation warnings as shown above.
terraform_validate ./examples/simple_zonal_private 
Success!
 The configuration is valid.
terraform_validate ./examples/simple_zonal_with_acm 
Success!
 The configuration is valid.
terraform_validate ./examples/simple_zonal_with_asm 
Success!
 The configuration is valid.
terraform_validate ./examples/simple_zonal_with_hub 
Success!
 The configuration is valid.
terraform_validate ./examples/simple_zonal_with_hub_kubeconfig 
Success!
 The configuration is valid.
terraform_validate ./examples/stub_domains 
Success!
 The configuration is valid.
terraform_validate ./examples/stub_domains_private 
Success!
 The configuration is valid.
terraform_validate ./examples/stub_domains_upstream_nameservers 
Success!
 The configuration is valid.
terraform_validate ./examples/upstream_nameservers 
Success!
 The configuration is valid.
terraform_validate ./examples/workload_identity 
Success!
 The configuration is valid.
terraform_validate ./examples/workload_metadata_config 
Success!
 The configuration is valid.
terraform_validate ./modules/acm 
Success!
 The configuration is valid.
terraform_validate ./modules/asm 
Success!
 The configuration is valid.
terraform_validate ./modules/auth 
Success!
 The configuration is valid.
terraform_validate ./modules/beta-private-cluster 
Success!
 The configuration is valid.
terraform_validate ./modules/beta-private-cluster-update-variant 
Success!
 The configuration is valid.
terraform_validate ./modules/beta-public-cluster 
Success!
 The configuration is valid.
terraform_validate ./modules/beta-public-cluster-update-variant 
Success!
 The configuration is valid.
terraform_validate ./modules/binary-authorization 
Success!
 The configuration is valid.
terraform_validate ./modules/config-sync 
Success!
 The configuration is valid.
terraform_validate ./modules/hub 
Success!
 The configuration is valid.
terraform_validate ./modules/k8s-operator-crd-support 
Success!
 The configuration is valid.
terraform_validate ./modules/private-cluster 
Success!
 The configuration is valid.
terraform_validate ./modules/private-cluster-update-variant 
Success!
 The configuration is valid.
terraform_validate ./modules/safer-cluster 
Success!
 The configuration is valid.
terraform_validate ./modules/safer-cluster-update-variant 
Success!
 The configuration is valid.
terraform_validate ./modules/services 
Success!
 The configuration is valid.
terraform_validate ./modules/workload-identity 
Success!
 The configuration is valid.
terraform_validate ./test/fixtures/beta_cluster 
Success!
 The configuration is valid.
terraform_validate ./test/fixtures/deploy_service 
Success!
 The configuration is valid.
terraform_validate ./test/fixtures/disable_client_cert 
Success!
 The configuration is valid.
terraform_validate ./test/fixtures/node_pool 
Success!
 The configuration is valid.
terraform_validate ./test/fixtures/node_pool_update_variant 
Success!
 The configuration is valid.
terraform_validate ./test/fixtures/private_zonal_with_networking 
Warning: 
Interpolation-only expressions are deprecated
  on ../../../examples/private_zonal_with_networking/main.tf line 42, in module "gcp-network":
  42:     
"${var.subnetwork}"
 = [
Terraform 0.11 and earlier required all non-constant expressions to be
provided via interpolation syntax, but this pattern is now deprecated. To
silence this warning, remove the "${ sequence from the start and the }"
sequence from the end of this expression, leaving just the inner expression.
Template interpolation syntax is still used to construct strings from
expressions when the template includes multiple interpolation sequences or a
mixture of literal strings and interpolations. This deprecation applies only
to templates that consist entirely of a single interpolation sequence.
Success!
 The configuration is valid, but there were some validation warnings as shown above.
terraform_validate ./test/fixtures/safer_cluster 
Warning: 
Interpolation-only expressions are deprecated
  on ../../../examples/safer_cluster/network.tf line 37, in module "gcp-network":
  37:     
"${local.subnet_name}"
 = [
Terraform 0.11 and earlier required all non-constant expressions to be
provided via interpolation syntax, but this pattern is now deprecated. To
silence this warning, remove the "${ sequence from the start and the }"
sequence from the end of this expression, leaving just the inner expression.
Template interpolation syntax is still used to construct strings from
expressions when the template includes multiple interpolation sequences or a
mixture of literal strings and interpolations. This deprecation applies only
to templates that consist entirely of a single interpolation sequence.
Success!
 The configuration is valid, but there were some validation warnings as shown above.
terraform_validate ./test/fixtures/safer_cluster_iap_bastion 
Warning: 
Interpolation-only expressions are deprecated
  on ../../../examples/safer_cluster_iap_bastion/network.tf line 36, in module "vpc":
  36:     
"${var.subnet_name}"
 = [
Terraform 0.11 and earlier required all non-constant expressions to be
provided via interpolation syntax, but this pattern is now deprecated. To
silence this warning, remove the "${ sequence from the start and the }"
sequence from the end of this expression, leaving just the inner expression.
Template interpolation syntax is still used to construct strings from
expressions when the template includes multiple interpolation sequences or a
mixture of literal strings and interpolations. This deprecation applies only
to templates that consist entirely of a single interpolation sequence.
Error: 
Unsupported argument
  on .terraform/modules/example.cloud-nat/main.tf line 58, in resource "google_compute_router_nat" "main":
  58:   
enable_endpoint_independent_mapping
 = var.enable_endpoint_independent_mapping
An argument named "enable_endpoint_independent_mapping" is not expected here.
terraform_validate ./test/fixtures/sandbox_enabled 
Success!
 The configuration is valid.
terraform_validate ./test/fixtures/simple_regional 
Success!
 The configuration is valid.
terraform_validate ./test/fixtures/simple_regional_private 
Success!
 The configuration is valid.
terraform_validate ./test/fixtures/simple_regional_with_kubeconfig 
Success!
 The configuration is valid.
terraform_validate ./test/fixtures/simple_regional_with_networking 
Warning: 
Interpolation-only expressions are deprecated
  on ../../../examples/simple_regional_with_networking/main.tf line 45, in module "gcp-network":
  45:     
"${var.subnetwork}"
 = [
Terraform 0.11 and earlier required all non-constant expressions to be
provided via interpolation syntax, but this pattern is now deprecated. To
silence this warning, remove the "${ sequence from the start and the }"
sequence from the end of this expression, leaving just the inner expression.
Template interpolation syntax is still used to construct strings from
expressions when the template includes multiple interpolation sequences or a
mixture of literal strings and interpolations. This deprecation applies only
to templates that consist entirely of a single interpolation sequence.
Success!
 The configuration is valid, but there were some validation warnings as shown above.
terraform_validate ./test/fixtures/simple_zonal 
Success!
 The configuration is valid.
terraform_validate ./test/fixtures/simple_zonal_private 
Success!
 The configuration is valid.
terraform_validate ./test/fixtures/simple_zonal_with_asm 
Success!
 The configuration is valid.
terraform_validate ./test/fixtures/stub_domains 
Success!
 The configuration is valid.
terraform_validate ./test/fixtures/stub_domains_private 
Success!
 The configuration is valid.
terraform_validate ./test/fixtures/stub_domains_upstream_nameservers 
Success!
 The configuration is valid.
terraform_validate ./test/fixtures/upstream_nameservers 
Success!
 The configuration is valid.
terraform_validate ./test/fixtures/workload_identity 
Success!
 The configuration is valid.
terraform_validate ./test/fixtures/workload_metadata_config 
Success!
 The configuration is valid.
terraform_validate ./test/setup 
Warning: 
Interpolation-only expressions are deprecated
  on .terraform/modules/gke-project-1/modules/project_services/main.tf line 57, in resource "google_project_iam_member" "project_service_identity_roles":
  57:   member  = "serviceAccount:${google_project_service_identity.project_service_identities[
"${each.value.api}"
].email}"
Terraform 0.11 and earlier required all non-constant expressions to be
provided via interpolation syntax, but this pattern is now deprecated. To
silence this warning, remove the "${ sequence from the start and the }"
sequence from the end of this expression, leaving just the inner expression.
Template interpolation syntax is still used to construct strings from
expressions when the template includes multiple interpolation sequences or a
mixture of literal strings and interpolations. This deprecation applies only
to templates that consist entirely of a single interpolation sequence.
(and 2 more similar warnings elsewhere)
Success!
 The configuration is valid, but there were some validation warnings as shown above.
 ```

kaariger commented:
I have committed two more changes:
- updated provider version in the safer_cluster_iap_bastion example to 2.52. The previous version was causing integration tests to fail.
- added a readme for version 14.0 since this is a breaking release. Added limitations around upgrades and installations for ASM versions.

bharathkkb commented on docs/upgrading_to_v14.0.md:
````diff
@@ -0,0 +1,24 @@
+[ASM submodule](https://github.com/terraform-google-modules/terraform-google-kubernetes-engine/tree/master/modules/asm) has been changed to use ASM v1.8 as default.
+
+```diff
+variable "asm_version" {
````
We could omit the diff in this case as the module user cannot use `1.6` with the new ASM module and maintain compatibility.

bharathkkb commented on docs/upgrading_to_v14.0.md:
```diff
@@ -0,0 +1,24 @@
+The v14.0 release of *kubernetes-engine* is a backwards incompatible
+release for some versions of Anthos Service Mesh (ASM).
+
+It uses the new ASM [installation script](https://cloud.google.com/service-mesh/docs/scripted-install/asm-onboarding) which:
```
We can move these down to the `### ASM default version changed to 1.8` heading
